physics/SFC_Models/Land/Noahmp/noahmpdrv.F90

```diff
@@ -9,99 +9,469 @@
 !! as diagnotics calculation.
 
 !> This module contains the CCPP-compliant NoahMP land surface model driver.
-      module noahmpdrv
+  module noahmpdrv
 
-      use module_sf_noahmplsm
+    use module_sf_noahmplsm
+    
+    ! Land IAU increments for soil temperature (can also do soil moisture increments if needed)
+    use land_iau_mod,  only: land_iau_control_type, land_iau_external_data_type,   &
+                            land_iau_mod_set_control, land_iau_mod_init,          &
+                            land_iau_mod_getiauforcing, land_iau_mod_finalize,    &
+                            calculate_landinc_mask
 
-      implicit none
+    implicit none
 
-      integer, parameter :: psi_opt = 0 ! 0: MYNN or 1:GFS
+    integer, parameter :: psi_opt = 0 ! 0: MYNN or 1:GFS
 
-      private
+    private
 
-      public :: noahmpdrv_init, noahmpdrv_run
+    public :: noahmpdrv_init, noahmpdrv_run, &
+              noahmpdrv_timestep_init, noahmpdrv_timestep_finalize, noahmpdrv_finalize
 
-      contains
+    !> \Land IAU data and control
+    ! Land IAU Control holds settings' information, maily read from namelist (e.g., 
+    ! block of global domain that belongs to a process ,
+    ! whethrer to do IAU increment at this time step, 
+    ! time step informatoin, etc)    
+    type (land_iau_control_type)          :: Land_IAU_Control
+    ! Land IAU Data holds spatially and temporally interpolated soil temperature increments per time step
+    type (land_iau_external_data_type)    :: Land_IAU_Data   !(number of blocks):each proc holds nblks
```

Comment:
@dustinswales Could you help me to understand whether this will cause an issue similar to https://github.com/ufs-community/ccpp-physics/issues/74?

Reply:
@dustinswales In your opinion, should this be done before this PR is approved/merged? I'm guessing that NRL is using NoahMP and therefore merging these changes would break their workflow? If so, we will probably need to provide some more guidance to move this along. I think that we need to come up with an implementation plan that we'd like to see followed. Also, I can't recall if we added something in the CCPP documentation about this issue, but we probably should so that we can refer to it and to help developers avoid this problem.

Reply:
We use NoahMP for testing. Can you help me refresh my memory what the problem is and if there is something I can do w.r.t. to testing?

Reply:
The issue is related to multiple instances of CCPP in one executable and the use of heap memory: https://github.com/ufs-community/ccpp-physics/issues/74

Reply:
See above. We have suites that use NoahMP for testing; the first operational implementation will use Noah.

If we can avoid breaking NoahMP for NEPTUNE, that would be great. What exactly is done here? Seeing this discussion/PR for the first time.

Reply:
Ok, I see it now. Those would have to be dimensioned as number of ccpp instances, or come from the host model

Reply:
@climbfuji The problem is related to data being stored on the HEAP when NEPTUNE has multiple ccpp instances in the same executable. A solution was to move these pieces from module memory and add them to the hosts typedefs, which I assume works for NEPTUNE? (We can't test this)

NoahMP has instances of its scheme-native DDTs stored on the HEAP. Will this cause a similar problem in NEPTUNE w/ multiple CCPP instances? I brought up RRTMGP because in there we use scheme-native DDTs, which are setup during initialization, so we are essentially doing the same thing as NoahMP, which may not work.
When NEPTUNE uses multiple instances, does each instance call ccpp init?

Reply:
Yes, each instance calls ccpp_init

Reply:
That's what I thought.
So if the writing to the DDTs was guarded by MPI directives, like in RRTMGP, we should be fine. No? (there is only one master processor...)

Reply:
For my understanding. Are these DDTs instance-dependent, i.e. are they potentially different for each instance, or are they identical regardless of the instance?

Reply:
In the case of the radiation, yes.
I can't speak for the LSM.

Reply:
This determines the approach we need to take. If the DDT is identical across all instances (e.g. containing static or time-dependent lookup tables etc.), then there can be one copy on the HEAP and initialization/destruction is done by the first instance but not the others.

If the DDT varies across instances, it either needs to be dimensioned by the number of CCPP instances, or it needs to be stored in the host model and passed as a variable.

Reply:
> In the case of the radiation, identical.
> I can't speak for the LSM.

@dustinswales different instances of the NoahMP LSM are likely to have different DDT values.

Reply:
Thanks @tsga, @climbfuji, @dustinswales for the discussion. I think that I understand the solution well enough to implement this here and in https://github.com/NOAA-EMC/fv3atm/pull/866. @tsga Would you be comfortable with me implementing this change and submitting PRs into [tsga:feature/lnd_iau](https://github.com/tsga/ccpp-physics/tree/feature/lnd_iau) and the associated fv3atm branch? Or, would you like to give it a shot?

Reply:
I guess that I should ask first, is there any reason why the instances of the types `land_iau_control_type, land_iau_external_data_type` can't be allocated in, e.g. GFS_typedefs.F90, and passed in? It's still fine to initialize them in the physics, but the memory just needs to "reside" in the host.

Reply:
@grantfirl, I can't see any reason against allocating them in the GFS_typedefs.F90.

Reply:
> Thanks @tsga, @climbfuji, @dustinswales for the discussion. I think that I understand the solution well enough to implement this here and in [NOAA-EMC/fv3atm#866](https://github.com/NOAA-EMC/fv3atm/pull/866). @tsga Would you be comfortable with me implementing this change and submitting PRs into [tsga:feature/lnd_iau](https://github.com/tsga/ccpp-physics/tree/feature/lnd_iau) and the associated fv3atm branch? Or, would you like to give it a shot?

@grantfirl, yes I would be comfortable with you implementing it. Thank you!

Reply:
@grantfirl Can you check if there is any performance impact when you are doing this? Just curious...